crates/goose/src/agents/chat_recall_extension.rs

```diff
@@ -0,0 +1,347 @@
+                before_date,
+                exclude_session_id,
+            )
+            .await
+            {
+                Ok(results) => {
+                    let formatted_results = if results.total_matches == 0 {
+                        format!("No results found for query: '{}'", query)
+                    } else {
+                        let mut output = format!(
+                            "Found {} matching message(s) across {} session(s) for query: '{}'\n\n",
+                            results.total_matches,
+                            results.results.len(),
+                            query
+                        );
+                        for (idx, result) in results.results.iter().enumerate() {
+                            output.push_str(&format!(
+                                "{}. Session: {} (ID: {})\n   Working Dir: {}\n   Last Activity: {}\n   Showing {} of {} total message(s) in session:\n\n",
+                                idx + 1,
+                                result.session_description,
+                                result.session_id,
+                                result.session_working_dir,
+                                result.last_activity.format("%Y-%m-%d"),
+                                result.messages.len(),
+                                result.total_messages_in_session
+                            ));
+
+                            for (msg_idx, message) in result.messages.iter().enumerate() {
+                                output.push_str(&format!(
+                                    "   {}.{} [{}]\n   {}\n\n",
+                                    idx + 1,
+                                    msg_idx + 1,
+                                    message.role,
+                                    message
+                                        .content
+                                        .lines()
+                                        .map(|line| format!("   {}", line))
+                                        .collect::<Vec<_>>()
+                                        .join("\n")
+                                ));
+                            }
+                        }
+                        output
+                    };
+                    Ok(vec![Content::text(formatted_results)])
+                }
+                Err(e) => Err(format!("Chat recall failed: {}", e)),
+            }
+        }
+    }
+
+    fn get_tools() -> Vec<Tool> {
+        vec![Tool::new(
+            "chat_recall".to_string(),
+            indoc! {r#"
+                Search past chat or load session summaries. Use when it is clear user expects some memory or context.
+
+                search mode (query): Use multiple keywords/synonyms. Returns messages grouped by session, ordered by recency. Supports date filters.
+                load mode (session_id): Returns first/last 3 messages of a session.
+            "#}
+            .to_string(),
+            object!({
```

Comment:
I know we do this in todo, but we shouldn't! handrolling of json schema is dangerous and will weirdly crash odd providers. Can you use the #[tool( macro instead /cc @alexhancock

Reply:
oh! I didn't know we could do this in non mcp-launched extensions (ie in process) - neat if so!